storybookjs/storybook
# CLI: Improve monorepo support in `automigrate` and revamp framework-related migrations

closes #20508, closes #20955

## What I did

### new CLI flags added to `storybook automigrate`

Automigrations now accept `--config-dir` that points to the `.storybook` directory that is to be transformed, and `--renderer` that has the renderer package of the project in use (e.g. @storybook/react, @storybook/angular). Both are optional. They are useful for monorepo-like scenarios, such as Nx, which will use these flags to migrate the existing projects that have storybook configuration.

Additionally, the `--skip-install` flag was also added.

For users in a monorepo that run the command without passing --config-dir, the migrations will stop with the following message:
<img width="623" alt="image" src="https://user-images.githubusercontent.com/1671563/219711089-c5429d24-ea93-405e-8405-0980aa946ca3.png">

The `renderer` flag is used to specify which framework the project is using. If this is not passed, and the workspace has multiple renderers installed, the user will get prompted:
<img width="436" alt="image" src="https://user-images.githubusercontent.com/1671563/219006512-2ed2deef-efb7-4d94-9a6e-358c09d04d93.png">

### Revamped automigrations

The following automigrations are now merged into a single one:
- mainjsFramework
- new-frameworks
- nextjs-framework
- sveltekit-framework

This way, users will get from zero to the correct point without having to run multiple migrations that tell them things like "you have a wrong setup, please upgrade to X. Well, actually.. you have a wrong setup, please upgrade to Y".

It's incremental, and there are dozens of possibilities to trigger the migration, where each step is shown like so:

![image](https://user-images.githubusercontent.com/1671563/219981176-0436a4dd-22c8-41cb-a625-7a5aa7064646.png)

The automigration also **does not** use AST parsing for _reading_ main.js, which makes them support non-standard main.js files like so:
```js
import getConfig from './baseConfig'
export default getConfig()
```

The automigration **will use** AST parsing for _writing_ into main.js when the user prompts that they want the migration to be done, and if there are failures, they will get a helpful message like so:
<img width="467" alt="image" src="https://user-images.githubusercontent.com/1671563/219011229-b098d2b4-9a52-49f9-879c-774e6e9394a7.png">

### Changes to AST tool: ConfigFile

Currently, the csf-tools helpers write a `named export` into a main.js file, even if it has `default export`. This is error prone and makes the end result incorrect. This PR makes that scenario throw an error instead.

### Misc

This PR adds a `jest_workaround` library to configure jest/swc given that the swc implementation has a limitation for mocking, which is solved by that dependency.

## TODO

- [x] Update MIGRATIONS.md

## How to test

There are lots of unit tests, please check them.

1. Go into a "migratable" Storybook project
2. Try to migrate it with `sb automigrate`
3. Try to migrate it with `sb automigrate` in a monorepo-like scenario (multiple renderers in package.json) and get prompted
4. Try to migrate it with `sb automigrate` in a monorepo-like scenario using the --renderer flag and do not get prompted
5. Try to migrate it with `sb automigrate` using --config-dir flag in a custom storybook config-dir scenario

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
